improve photometry plot

dannygoldstein opened the pull request:
- plot in both flux and magnitude space
- show flux/mag errors on plot
- x axis as mjd instead of timestamp
- slider bar to stack light curve points

dannygoldstein commented:
<img width="828" alt="Screen Shot 2019-06-19 at 10 00 01 PM" src="https://user-images.githubusercontent.com/2769632/59819642-a24e9f00-92dd-11e9-97e9-e6747f72b501.png">
<img width="839" alt="Screen Shot 2019-06-19 at 10 00 12 PM" src="https://user-images.githubusercontent.com/2769632/59819643-a5498f80-92dd-11e9-9db8-33d12c8a8ac7.png">

stefanv commented:
Thanks, @dannygoldstein. The plots look really nice.

This PR introduces dependencies on astropy and sncosmo.  Are both strictly necessary?

dannygoldstein commented:
Yes. `sncosmo` handles normalization to a common flux system of photometric data points from images with different zeropoints and magnitude systems. astropy==3.0.3 is already a requirement, introduced in this commit:

 2ca3189d (Ari Crellin-Quick   2018-06-21 17:51:18 -0700  9) astropy==3.0.3).


This PR is ready for review.

acrellin commented:
@dannygoldstein thanks for the PR!  I see that Travis is erroring on the new DB changes; do you want a hand getting that sorted out?

acrellin commented:
@dannygoldstein what's the status of this PR? Do you want a hand getting this working?

acrellin commented:
@dannygoldstein what's the status of this PR? Should we try to get this working, or close it?

acrellin commented:
@dannygoldstein can you move the Javascript files to static/js?

dannygoldstein commented:
Hi all, this PR now makes an important change - 

https://github.com/skyportal/skyportal/pull/191/commits/f8487ebb8e7f862b897abc01774a0af9824c8653#diff-d5d626afdf1778c3daae39428ee3ac4dR273

To handle accepting magnitude systems in the photometry class, and to handle forced photometry data in flux space (which is necessary for stacking photometry), we need an implementation of each valid magnitude system, as well as a list of allowed bandpasses and their measured throughputs. 

In this PR I have defined a set of allowed bandpasses and magnitude systems by pegging what is allowed for these values to what is defined in sncosmo, which implements these magnitude systems:

https://sncosmo.readthedocs.io/en/latest/magsystem-list.html

 and defines (and versions) the following bandpasses 

https://sncosmo.readthedocs.io/en/latest/bandpass-list.html

The downside is that this fixes us to only accept photometry in specific bandpasses, but the list is pretty large and it is fairly easy to  add new ones. 

Requesting @profjsb  comments on this

dannygoldstein commented:
> @dannygoldstein can you move the Javascript files to static/js?

Done.

acrellin commented:
@dannygoldstein `make load_demo_data` fails for me locally when I try to run this branch, with errors related to these enum fields.

dannygoldstein commented:
Yes so the error is

```
sqlalchemy.exc.StatementError: (builtins.LookupError) "g" is not among the defined enum values [SQL: 'INSERT INTO photometry (created_at, observed_at, mjd, time_format, time_scale, mag, e_mag, lim_mag, flux, fluxerr, zp, zpsys, ra, dec, filter, isdiffpos, var_mag, var_e_mag, dist_nearest_source, mag_nearest_source, e_mag_nearest_source, score, candid, origin, source_id, instrument_id) VALUES (%(created_at)s, %(observed_at)s, %(mjd)s, %(time_format)s, %(time_scale)s, %(mag)s, %(e_mag)s, %(lim_mag)s, %(flux)s, %(fluxerr)s, %(zp)s, %(zpsys)s, %(ra)s, %(dec)s, %(filter)s, %(isdiffpos)s, %(var_mag)s, %(var_e_mag)s, %(dist_nearest_source)s, %(mag_nearest_source)s, %(e_mag_nearest_source)s, %(score)s, %(candid)s, %(origin)s, %(source_id)s, %(instrument_id)s) RETURNING photometry.id'] [parameters: [{'source_id': '14gqr', 'instrument_id': 1, 'mag': 99.0, 'lim_mag': 21.7691, 'e_mag': 99.0, 'filter': 'g', 'time_scale': 'tcb', 'observed_at': '2014-09-01 06:49:40.013', 'time_format': 'iso', 'zp': None, 'zpsys': None, 'origin': None, 'dec': None, 'candid': None, 'ra': None, 'e_mag_nearest_source': None, 'fluxerr': None, 'var_e_mag': None, 'mjd': None, 'mag_nearest_source': None, 'flux': None, 'score': None, 'var_mag': None, 'dist_nearest_source': None}]]
```

Which is because "g" is not a well defined filter -- there are many different "g" filters out there. With this new fix the filter would need to be called "ztfg" (if it referring to data from ZTF) or "sdssg" (if it is referring to data from SDSS) or "desg" (if it is referring to data from DECam)

acrellin commented:
If you want to update that (and anywhere else those filter values are used,
e.g. instrument_observation_params.json) go ahead and replace them with
whichever of those you'd like (doesn't matter for demo data purposes).

On Sat, May 2, 2020 at 12:37 PM Danny Goldstein <notifications@github.com>
wrote:

> Yes so the error is
>
> sqlalchemy.exc.StatementError: (builtins.LookupError) "g" is not among the defined enum values [SQL: 'INSERT INTO photometry (created_at, observed_at, mjd, time_format, time_scale, mag, e_mag, lim_mag, flux, fluxerr, zp, zpsys, ra, dec, filter, isdiffpos, var_mag, var_e_mag, dist_nearest_source, mag_nearest_source, e_mag_nearest_source, score, candid, origin, source_id, instrument_id) VALUES (%(created_at)s, %(observed_at)s, %(mjd)s, %(time_format)s, %(time_scale)s, %(mag)s, %(e_mag)s, %(lim_mag)s, %(flux)s, %(fluxerr)s, %(zp)s, %(zpsys)s, %(ra)s, %(dec)s, %(filter)s, %(isdiffpos)s, %(var_mag)s, %(var_e_mag)s, %(dist_nearest_source)s, %(mag_nearest_source)s, %(e_mag_nearest_source)s, %(score)s, %(candid)s, %(origin)s, %(source_id)s, %(instrument_id)s) RETURNING photometry.id'] [parameters: [{'source_id': '14gqr', 'instrument_id': 1, 'mag': 99.0, 'lim_mag': 21.7691, 'e_mag': 99.0, 'filter': 'g', 'time_scale': 'tcb', 'observed_at': '2014-09-01 06:49:40.013', 'time_format': 'iso', 'zp': None, 'zpsys': None, 'origin': None, 'dec': None, 'candid': None, 'ra': None, 'e_mag_nearest_source': None, 'fluxerr': None, 'var_e_mag': None, 'mjd': None, 'mag_nearest_source': None, 'flux': None, 'score': None, 'var_mag': None, 'dist_nearest_source': None}]]
>
> Which is because "g" is not a well defined filter -- there are many
> different "g" filters out there. With this new fix the filter would need to
> be called "ztfg" (if it referring to data from ZTF) or "sdssg" (if it is
> referring to data from SDSS) or "desg" (if it is referring to data from
> DECam)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-623003381>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTOFJXTF4XH75LHYTXDRPRZA7ANCNFSM4HZPKXDA>
> .
>

dannygoldstein commented:
Fixed. @acrellin one thing that would be really helpful would be if you could lint the js?

acrellin commented:
@dannygoldstein sure, happy to help with the linter errors. Is the photometry plot supposed to be working as-is? I'm getting errors, and the plot doesn't load: 
```
ERROR:tornado.application:Uncaught exception GET /api/internal/plot/photometry/14gqr (127.0.0.1)          
HTTPServerRequest(protocol='http', host='localhost:5000', method='GET', uri='/api/internal/plot/photometry
/14gqr', version='HTTP/1.0', remote_ip='127.0.0.1')                                                       
Traceback (most recent call last):                                                                        
  File "/home/arien/anaconda3/lib/python3.7/site-packages/tornado/web.py", line 1697, in _execute         
    result = method(*self.path_args, **self.path_kwargs)                                                  
  File "/home/arien/projects/skyportal/baselayer/app/access.py", line 31, in wrapper                      
    return tornado.web.authenticated(method)(self, *args, **kwargs)                                       
  File "/home/arien/anaconda3/lib/python3.7/site-packages/tornado/web.py", line 3174, in wrapper          
    return method(self, *args, **kwargs)                                                                  
  File "/home/arien/projects/skyportal/skyportal/handlers/api/internal/plot.py", line 10, in get          
    docs_json, render_items, custom_model_js = plot.photometry_plot(source_id)                            
  File "/home/arien/projects/skyportal/skyportal/plot.py", line 180, in photometry_plot                   
    normalized = photdata.normalized(zp=25., zpsys='ab')                                                  
  File "/home/arien/anaconda3/lib/python3.7/site-packages/sncosmo/photdata.py", line 128, in normalized   
    factor = self._normalization_factor(zp, zpsys)                                                        
  File "/home/arien/anaconda3/lib/python3.7/site-packages/sncosmo/photdata.py", line 152, in _normalizatio
n_factor                                                                                                  
    bandfactor = 10.**(0.4 * (zp - self.zp[mask]))                                                        
TypeError: unsupported operand type(s) for -: 'float' and 'NoneType'                                      
! App Error: unsupported operand type(s) for -: 'float' and 'NoneType'                                    
WARNING:tornado.access:400 GET /api/internal/plot/photometry/14gqr (127.0.0.1) 60.45ms                    
```

acrellin commented:
@dannygoldstein https://github.com/dannygoldstein/skyportal/pull/1  fixes the linter errors.

dannygoldstein commented:
Thanks @acrellin , just have to make one change to accommodate upper limits with no corresponding flux measurements

profjsb commented:
@dannygoldstein can you describe how photometry from “negative” subtractions will be loaded? We clearly need to do some math/coercion at the api level before the DB is stuffed with fluxes. I don’t see tests for negative subtractions...this is important for varstars found in difference imaging.

dannygoldstein commented:
Negative fluxes are handled naturally by this approach and render as negative fluxes (normalized to a common zeropoint) on the flux plot and an  upper limit on the magnitude plot (we do not assume a baseline flux for the varstars).

Example:

<img width="602" alt="Screen Shot 2020-05-02 at 5 05 30 PM" src="https://user-images.githubusercontent.com/2769632/80892339-3b3c9d80-8c97-11ea-9185-189e63794431.png">
<img width="635" alt="Screen Shot 2020-05-02 at 5 05 34 PM" src="https://user-images.githubusercontent.com/2769632/80892341-3d066100-8c97-11ea-97d6-8f6a9a26fe45.png">
<img width="366" alt="Screen Shot 2020-05-02 at 5 05 52 PM" src="https://user-images.githubusercontent.com/2769632/80892342-3d066100-8c97-11ea-9a40-0812ef997cb6.png">

dannygoldstein commented:
So @profjsb the negative flux checking is effectively happening here:

https://github.com/dannygoldstein/skyportal/blob/ff7f5092c1447038bde681511f088332a361b5e5/skyportal/plot.py#L191

dmitryduev commented:
@dannygoldstein @acrellin Wanted to mention #415 here. Is this something that should simple be defined in instrument_observation_params.json per instrument per filter?

dannygoldstein commented:
@dmitryduev yeah. we can set sensible defaults. the currently defined bandpasses are

'f1550c', 'swope2::v1', 'standard::b', 'keplercam::b', 'f210m', 'ztfr', 'f555w', 'csphs', 'f300x', 'swope2::v', 'f625w', 'cspr', 'uvf555w', 'cspg', 'keplercam::r', 'keplercam::v', 'lsstu', 'swope2::b', 'cspv9844', 'swope2::g', 'sdss::i', 'lsstr', 'bessellv', 'uvf606w', 'f139m', 'sdssr', 'uvf775w', 'desy', 'f480m', 'f2550w', 'ztfi', 'f606w', 'f125w', 'f140w', 'f277w', 'f110w', 'f689m', 'cspk', 'nicmos2::f110w', 'cspu', 'f1140c', 'swope2::i', 'bessellb', 'f140m', 'f763m', 'f2100w', 'acswf::f775w', 'swope2::r', 'f775w', 'f430m', 'f300m', 'swope2::v2', 'f770w', 'swope2::y', 'f2300c', 'f460m', '4shooter2::b', 'f1500w', 'f1000w', 'f390w', 'swope2::u', 'f1280w', 'standard::r', 'sdssi', 'uvf475w', 'f435w', 'swope2::h', 'sdss::u', 'f475w', 'nicmos2::f160w', 'f275w', 'desr', 'ztfg', 'lssti', 'lsstz', 'f350lp', 'nicf160w', '4shooter2::i', 'lssty', 'uvf850lp', 'f845m', 'f200w', 'keplercam::i', 'cspi', 'f560w', 'keplercam::us', 'f162m', 'f1065c', 'standard::v', 'sdssg', 'f850lp', 'cspb', 'acswf::f606w', 'cspjd', 'desi', 'f090w', 'sdssz', '4shooter2::us', 'sdss::r', 'f444w', 'standard::u', 'f182m', 'f1130w', 'kepler', 'cspv3014', 'besselli', 'cspys', 'bessellr', 'f410m', 'standard::i', 'lsstg', 'f336w', 'f105w', 'uvf814w', 'sdss::z', 'f160w', 'f098m', 'f438w', 'f153m', 'cspjs', 'f150w', 'cspv3009', 'uvf625w', 'f1800w', 'f356w', 'f250m', '4shooter2::v', 'desz', 'swope2::j', 'desg', 'csphd', 'sdss::g', 'f115w', '4shooter2::r', 'nicf110w', 'f335m', 'acswf::f850lp', 'bessellux', 'f225w', 'sdssu', 'cspyd', 'f127m', 'f070w', 'f360m', 'f218w'

dannygoldstein commented:
Another issue here. 

In the case of a non detection, the current official ZTF alerts only provide an upper limit derived over the entire image. No photometry is performed at the location of the object, so no flux is measured when nothing is detected.

This is equivalent to specifying a magnitude error or a flux error in the case of a non detection, but not an actual flux or magnitude. 

Without flux measurements for everything (including non detections), it is not totally clear how the light curve stacking should work. Stacking happens in flux space, and needs flux and flux error measurements for every point in order to combine them. For ZUDS, which is already using skyportal and the changes proposed on this PR, we are already doing forced photometry for all images containing a source, including non detections, so it has been possible to do light curve stacking. Performing forced photometry to derive flux measurements for non detections in the  alerts is currently in discussion for ZTF 2. 

One solution would be to simply exclude limits with no corresponding flux measurement from the stack, however this could be a little misleading. People will think they are combining limits to achieve a deeper mag limit, when in reality most of the limits are being left as-is because there is no way to combine them.

Another would be to mandate that our model of photometry be flux, fluxerror, time, filter, zp, and zpsys (as in sncosmo), so flux and flux error measurements would be required for every photometry point. This would solve the problem entirely, but would preclude ingestion of ZTF1 alerts with no measured flux where there is an upper limit.

dannygoldstein commented:
cc @doccosmos

dannygoldstein commented:
Hi @acrellin, @stefanv, @profjsb this is passing tests and ready for review. The only failing tests are related to Selenium and are fixed by @acrellin on master. This PR should also supersede #430 (it includes the commits from that PR and also a few important updates).

dannygoldstein commented:
I added a few more changes - core functionality here is done now from my end. 

   * Added the ability to display magnitude-only upper limits (i.e., ZTF1 alert upper limits) and incorporate them self-consistently into the light curve stacking
   * Added a new colormap for photometry plot based on bandpass effective wavelength
   * Added some more validation for photometry endpoint based on new database constraints for new photometry data model
   * Made sure the export LC button exports both stacked and unstacked data when binsize > 0 and some data has no measured flux (i.e., ZTF1 upper limits).

dannygoldstein commented:
The approach I took to solve the problem described in the long essay above is to basically leave the flux-less upper limits "as is" in the stacking and add a field to the hover tooltip that says "stacked=True" when a point comes from light curve stacking and "stacked=False" when something cannot be stacked due to lack of flux information. If a single-epoch measurement is used in a stacked point then its alpha is reduced

dannygoldstein commented:
@acrellin I am unable to reproduce the failures on travis locally, I think they are all related to selenium, the only error I can produce locally is on `test_news_feed_prefs_widget` which fails with `/anaconda2/envs/skyportal/lib/python3.6/site-packages/selenium/webdriver/remote/errorhandler.py:242: ElementClickInterceptedException`, and is different from the two travis failures

acrellin commented on static/js/plotjs/stackm.js:
```diff
@@ -3,146 +3,203 @@ const binsize = slider.value;
+
+    if (binsize === 0) {
+        for (var j = 0; j < allsource.get_length(); j++) {
+            boldsource.data.flux.push(allsource.data.flux[j]);
```
This can be simplified by iterating over the list of relevant keys (fine leaving as is for now). The other section below, as well.

acrellin commented:
@dannygoldstein hmm, these all pass for me locally, but it's likely just due to differences in screen resolution/browser window size. I'll push up a fix for these

stefanv commented:
Nice work!  @dannygoldstein Did you want to do any JS cleanups still, or shall we merge and get back to it later?

dannygoldstein commented:
I vote merge now and we can clean up in advance of first release

On Sat, May 9, 2020 at 3:10 PM Stefan van der Walt <notifications@github.com>
wrote:

> Nice work! @dannygoldstein <https://github.com/dannygoldstein> Did you
> want to do any JS cleanups still, or shall we merge and get back to it
> later?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626222309>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVEFYEHS4NDRCQSSXYLMOTRQWTBXANCNFSM4HZPKXDA>
> .
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

dannygoldstein commented:
Although it may make sense to write some tests using selenium... but I dont
really know how to do that. Happy to wait on that if you think it is
necessary

On Sat, May 9, 2020 at 3:19 PM danny goldstein <dgold@berkeley.edu> wrote:

> I vote merge now and we can clean up in advance of first release
>
> On Sat, May 9, 2020 at 3:10 PM Stefan van der Walt <
> notifications@github.com> wrote:
>
>> Nice work! @dannygoldstein <https://github.com/dannygoldstein> Did you
>> want to do any JS cleanups still, or shall we merge and get back to it
>> later?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626222309>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAVEFYEHS4NDRCQSSXYLMOTRQWTBXANCNFSM4HZPKXDA>
>> .
>>
>
>
> --
> Danny Goldstein
> http://astro.caltech.edu/~danny
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

acrellin commented:
@dannygoldstein to use selenium for moving the slider, you can do something like this: 
```
from selenium.webdriver.common.action_chains import ActionChains


move = ActionChains(driver)
move.click_and_hold(slider).move_by_offset(40, 0).release().perform()
```

dannygoldstein commented:
where is `slider` initialized?

On Sun, May 10, 2020 at 1:40 PM Ari Crellin-Quick <notifications@github.com>
wrote:

> @dannygoldstein <https://github.com/dannygoldstein> to use selenium for
> moving the slider, you can do something like this:
>
> from selenium.webdriver.common.action_chains import ActionChains
>
>
> move = ActionChains(driver)
> move.click_and_hold(slider).move_by_offset(40, 0).release().perform()
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626363005>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVEFYFK37INEPCBXNJXFHDRQ3RJRANCNFSM4HZPKXDA>
> .
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

acrellin commented:
See edited comment

On Sun, May 10, 2020, 1:17 PM Danny Goldstein <notifications@github.com>
wrote:

> where is `slider` initialized?
>
> On Sun, May 10, 2020 at 1:40 PM Ari Crellin-Quick <
> notifications@github.com>
> wrote:
>
> > @dannygoldstein <https://github.com/dannygoldstein> to use selenium for
> > moving the slider, you can do something like this:
> >
> > from selenium.webdriver.common.action_chains import ActionChains
> >
> >
> > move = ActionChains(driver)
> > move.click_and_hold(slider).move_by_offset(40, 0).release().perform()
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/skyportal/skyportal/pull/191#issuecomment-626363005
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAVEFYFK37INEPCBXNJXFHDRQ3RJRANCNFSM4HZPKXDA
> >
> > .
> >
>
>
> --
> Danny Goldstein
> http://astro.caltech.edu/~danny
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626383450>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTKMBMGANFN4UYVX2W3RQ4DVVANCNFSM4HZPKXDA>
> .
>

profjsb commented:
I'm still not able to test this branch. First doing:
```
make db_init
make load_demo_data
```
Then doing a hard refresh on the browser (both Chrome and Firefox). In the log I get:
```
[service/webpack] Hash: fd9ab8825cb530f1a416
[service/webpack] Version: webpack 4.43.0
[service/webpack] Time: 739ms
[service/webpack] Built at: 05/10/2020 2:02:56 PM
[service/webpack]     Asset     Size  Chunks             Chunk Names
[service/webpack] bundle.js  571 KiB    main  [emitted]  main
[service/webpack] Entrypoint main = bundle.js
[service/webpack] [0] multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx 52 bytes {main} [built]
[service/webpack] [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {main} [built]
[service/webpack]     + 332 hidden modules
[service/webpack] 
[service/webpack] ERROR in multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx
[service/webpack] Module not found: Error: Can't resolve '/Users/.../Projects/skyportal/static/js/components/Main.jsx' in '/Users/.../Projects/skyportal'
[service/webpack]  @ multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx main[2]
```
And in the javascript console I get:
```
Main.jsx:3 Uncaught Error: Cannot find module '/Users/jbloom/Projects/skyportal/static/js/components/Main.jsx'
    at webpackMissingModule (Main.jsx:3)
    at eval (Main.jsx:3)
    at Object.0 (bundle.js:3860)
    at __webpack_require__ (bundle.js:20)
    at bundle.js:84
    at bundle.js:87
```

acrellin commented:
Ah, Josh, you may need to fetch a newer baselayer (git submodule update)

On Sun, May 10, 2020, 2:06 PM Joshua Bloom <notifications@github.com> wrote:

> I'm still not able to test this branch. First doing:
>
> make db_init
> make load_demo_data
>
> Then doing a hard refresh on the browser (both Chrome and Firefox). In the
> log I get:
>
> [service/webpack] Hash: fd9ab8825cb530f1a416
> [service/webpack] Version: webpack 4.43.0
> [service/webpack] Time: 739ms
> [service/webpack] Built at: 05/10/2020 2:02:56 PM
> [service/webpack]     Asset     Size  Chunks             Chunk Names
> [service/webpack] bundle.js  571 KiB    main  [emitted]  main
> [service/webpack] Entrypoint main = bundle.js
> [service/webpack] [0] multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx 52 bytes {main} [built]
> [service/webpack] [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {main} [built]
> [service/webpack]     + 332 hidden modules
> [service/webpack]
> [service/webpack] ERROR in multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx
> [service/webpack] Module not found: Error: Can't resolve '/Users/.../Projects/skyportal/static/js/components/Main.jsx' in '/Users/.../Projects/skyportal'
> [service/webpack]  @ multi whatwg-fetch babel-polyfill ./static/js/components/Main.jsx main[2]
>
> And in the javascript console I get:
>
> Main.jsx:3 Uncaught Error: Cannot find module '/Users/jbloom/Projects/skyportal/static/js/components/Main.jsx'
>     at webpackMissingModule (Main.jsx:3)
>     at eval (Main.jsx:3)
>     at Object.0 (bundle.js:3860)
>     at __webpack_require__ (bundle.js:20)
>     at bundle.js:84
>     at bundle.js:87
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626389650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTNA6BYZ5NTFX7PJMDTRQ4JLRANCNFSM4HZPKXDA>
> .
>

profjsb commented:
That helped, @acrellin. `make load_demo_data` starts fine, but after the homescreen renders, it looks like the system fails pretty quickly:
```
....
2020/05/10 15:02:21 [warn] 67878#0: *57 a client request body is buffered to a temporary file ./tmp/client_body/0000000013, client: 127.0.0.1, server: , request: "POST /api/thumbnail HTTP/1.1", host: "localhost:5000"
2020/05/10 15:02:23 [notice] 67874#0: signal 15 (SIGTERM) received from 67866, exiting
2020/05/10 15:02:23 [notice] 67878#0: exiting
2020/05/10 15:02:23 [notice] 67874#0: signal 23 (SIGIO) received from 67878
2020/05/10 15:02:23 [notice] 67878#0: exit
2020/05/10 15:02:23 [notice] 67874#0: signal 20 (SIGCHLD) received from 67878
2020/05/10 15:02:23 [notice] 67874#0: worker process 67878 exited with code 0
2020/05/10 15:02:23 [notice] 67874#0: exit
```

acrellin commented:
That's the log output for make run?

On Sun, May 10, 2020, 3:04 PM Joshua Bloom <notifications@github.com> wrote:

> That helped, @acrellin <https://github.com/acrellin>. make load_demo_data
> starts fine, but after the homescreen renders, it looks like the system
> fails pretty quickly:
>
> ....
> 2020/05/10 15:02:21 [warn] 67878#0: *57 a client request body is buffered to a temporary file ./tmp/client_body/0000000013, client: 127.0.0.1, server: , request: "POST /api/thumbnail HTTP/1.1", host: "localhost:5000"
> 2020/05/10 15:02:23 [notice] 67874#0: signal 15 (SIGTERM) received from 67866, exiting
> 2020/05/10 15:02:23 [notice] 67878#0: exiting
> 2020/05/10 15:02:23 [notice] 67874#0: signal 23 (SIGIO) received from 67878
> 2020/05/10 15:02:23 [notice] 67878#0: exit
> 2020/05/10 15:02:23 [notice] 67874#0: signal 20 (SIGCHLD) received from 67878
> 2020/05/10 15:02:23 [notice] 67874#0: worker process 67878 exited with code 0
> 2020/05/10 15:02:23 [notice] 67874#0: exit
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626396694>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABXFGTJZLKC47UXU7WWUENDRQ4QHLANCNFSM4HZPKXDA>
> .
>

dannygoldstein commented:
Hi Josh,

I think make load_demo_data is supposed to just load the demo data and die
that way (see the finally block here):

https://github.com/skyportal/skyportal/blob/master/tools/load_demo_data.py#L261

After you run make load_demo_data, you have to do make run to actually run
the app



On Sun, May 10, 2020 at 6:48 PM Ari Crellin-Quick <notifications@github.com>
wrote:

> That's the log output for make run?
>
> On Sun, May 10, 2020, 3:04 PM Joshua Bloom <notifications@github.com>
> wrote:
>
> > That helped, @acrellin <https://github.com/acrellin>. make
> load_demo_data
> > starts fine, but after the homescreen renders, it looks like the system
> > fails pretty quickly:
> >
> > ....
> > 2020/05/10 15:02:21 [warn] 67878#0: *57 a client request body is
> buffered to a temporary file ./tmp/client_body/0000000013, client:
> 127.0.0.1, server: , request: "POST /api/thumbnail HTTP/1.1", host:
> "localhost:5000"
> > 2020/05/10 15:02:23 [notice] 67874#0: signal 15 (SIGTERM) received from
> 67866, exiting
> > 2020/05/10 15:02:23 [notice] 67878#0: exiting
> > 2020/05/10 15:02:23 [notice] 67874#0: signal 23 (SIGIO) received from
> 67878
> > 2020/05/10 15:02:23 [notice] 67878#0: exit
> > 2020/05/10 15:02:23 [notice] 67874#0: signal 20 (SIGCHLD) received from
> 67878
> > 2020/05/10 15:02:23 [notice] 67874#0: worker process 67878 exited with
> code 0
> > 2020/05/10 15:02:23 [notice] 67874#0: exit
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/skyportal/skyportal/pull/191#issuecomment-626396694
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABXFGTJZLKC47UXU7WWUENDRQ4QHLANCNFSM4HZPKXDA
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skyportal/skyportal/pull/191#issuecomment-626401627>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVEFYCLADXAZKJQB3YCBDLRQ4VMPANCNFSM4HZPKXDA>
> .
>


-- 
Danny Goldstein
http://astro.caltech.edu/~danny

profjsb commented on skyportal/handlers/api/photometry.py:
```diff
@@ -38,44 +43,47 @@ def post(self):
-            data['e_mag'] = [data['e_mag']]
-            data['lim_mag'] = [data['lim_mag']]
-            data['filter'] = [data['filter']]
+        if 'mjd' not in data :
```
```suggestion
        if 'mjd' not in data:
```

profjsb commented on skyportal/handlers/api/photometry.py:
```diff
@@ -15,6 +10,16 @@
 
 
+PHOTOMETRY_COLUMNS = ['mjd', 'flux', 'fluxerr', 'zpsys', 'zp', 'filter']
+
```
```suggestion


```
PEP8: Two spaces before function def.

dannygoldstein commented:
@profjsb the meaning of 'Flux (AB Zeropoint = 25)' is that if you take 

  -2.5 log10(flux) + 25, 

you get an AB magnitude. The reason the zoomed-out flux plot does not show all of the data (limits) in the mag plot is that many of the limits in the mag plot do not have an associated flux measurement (in the input data, they are given as magnitude upper limits only -- i.e., e_mag = Null, mag = Null, maglimit = something).

profjsb commented on skyportal/models.py:
```diff
@@ -283,6 +296,40 @@ class Photometry(Base):
+    def mag(cls):
+        return sa.case(
+            [
+                (sa.and_(cls.flux > 0, cls.zp != None),
```
```suggestion
                (sa.and_(cls.flux > 0, cls.zp is not None),
```

dannygoldstein replied:
Peculiarity of SQLalchemy, need to write this as `!=`.

acrellin replied:
typically I've seen `colname.isnot(None)` recommended

profjsb commented on static/js/components/ThumbnailList.jsx:
```diff
@@ -10,7 +10,9 @@ import styles from "./ThumbnailList.css";
 
-const Thumbnail = ({ ra, dec, telescope, observed_at, name, url }) => {
+const Thumbnail = ({ ra, dec, telescope, mjd, name, url }) => {
+  const unixt = (mjd - 40587.0) * 86400000;
```
Add a small comment to document where 40587.0 comes from ... assuming it's an mjd to unix time conversion? What does the whole equation give you?

dannygoldstein replied:
The whole equation converts the given mjd into a unix timestap in ms

dannygoldstein replied:
40587 is the MJD of the last UNIX timestamp epoch (1970-01-01)

profjsb commented on static/js/plotjs/download.js:
```diff
@@ -0,0 +1,53 @@
+}
+
+
+const filename = 'objname.csv';
```
can this file name be dynamically created based on source position or source name? I'm worried about having a bunch of csv files with named `objname (N).csv` in the Downloads folder.

dannygoldstein replied:
This string is just a placeholder and is automatically replaced by the object name in python

profjsb commented on static/js/plotjs/download.js:
```diff
@@ -0,0 +1,53 @@
+  const lines = [];
+
+  if (write_header) {
+    lines.push(columns.join(','));
```
Add a comment in the header with filename, download time, etc. Like
```pre
# source: ZUDS1234 downloaded at: 2020-05-12T12:12:30
mjd, filter, ....
```

profjsb commented on static/js/plotjs/stackf.js:
```diff
@@ -0,0 +1,115 @@
+  const fluxerrsource = eval(`obserr${i}`).data_source;
+  const binerrsource = eval(`binerr${i}`).data_source;
+
+  const minmjd = Math.min.apply(Math, fluxsource.data.mjd) - 15;
```
See previous review comment on the suggested zoom windows for t and flux/mag.

dannygoldstein replied:
This is just for stacking windows, not for zoom windows

profjsb commented on static/js/plotjs/stackf.js:
```diff
@@ -0,0 +1,115 @@
+      const myfluxerr = Math.sqrt(1 / ivarsum);
+
+
+      if (myflux / myfluxerr > 5.0) {
```
Do you want to hardcode this threshold or make it changeable to the end user on the front end (or a config file)?

profjsb commented on static/js/plotjs/stackm.js:
```diff
@@ -0,0 +1,205 @@
+            }
+
+            const myfluxerr = Math.sqrt(1 / ivarsum);
+            const obs = myflux / myfluxerr > 5;
```
See question about about this threshold.

profjsb commented:
> @profjsb the meaning of 'Flux (AB Zeropoint = 25)' is that if you take
> 
> -2.5 log10(flux) + 25,
> 
> you get an AB magnitude. The reason the zoomed-out flux plot does not show all of the data (limits) in the mag plot is that many of the limits in the mag plot do not have an associated flux measurement (in the input data, they are given as magnitude upper limits only -- i.e., e_mag = Null, mag = Null, maglimit = something).

AB mag is just a conversion to/from cgs. But the way it's presented is that Flux (f_nu? or f_lam?) is dimensionless, which it's not. I'd prefer "Flux (muJy)" and then export to CSV however you like.

stefanv commented:
Thank you for the thorough review, @profjsb.

dannygoldstein commented:
@profjsb waiting to add date to hover tip for  #307

dannygoldstein commented:
> * On the flux plot: The units of Flux is not "AB," but rather something like $\mu$Jy. Doesn't make sense to say zeropoint = 25.0 if you've normalized the mags to a common (cgs) flux.

Added $\mu$Jy.

> * I think a better original zoom is to use a slightly wider y-axis scale than the faintest and brightest _detections_ (rather than limits). So [limits] +/- 0.1 mag (mag) and [limits] *1.05 /0.95 (flux).

Didn't implement this as it has undefined behavior in the case where there are no detections of a source (forced photometry). 

> * Make sure that bokeh sends over ALL the data so when you zoom out that data is still there.

See comment above about flux vs magnitude limits / detections. Can't plot magnitude upper limits in flux space. 

> 
> Some UI updates/changes:
> ![Screen_Shot_2020-05-10_at_3_10_44_PM](https://user-images.githubusercontent.com/480799/81595523-79ad1900-9377-11ea-9ffa-5a3af933dea3.jpg)
> 
> ![Screen_Shot_2020-05-10_at_3_08_41_PM](https://user-images.githubusercontent.com/480799/81596509-145a2780-9379-11ea-8a78-03c19ccbe613.jpg)
> ![Screen_Shot_2020-05-10_at_3_07_32_PM](https://user-images.githubusercontent.com/480799/81596757-7581fb00-9379-11ea-9b99-f46c1403009c.jpg)

acrellin commented:
@dannygoldstein do you mind addressing these merge conflicts? If you don't have bandwidth right now I can do it. Then we'll be good to go.

stefanv commented:
@dannygoldstein As per the meeting today, do you mind filing issues for the items that will not be addressed during this PR but that we're raised during review?  Thanks very much!

dannygoldstein commented:
Build is failing due to #456 but once NERSC comes back up the tests should pass. This is ready to merge so I am closing and merging. Future PRs will extend this PR with:

 * further improvements to Photometry storage and data model
 * aesthetic improvements described above by @profjsb

dannygoldstein commented:
.

acrellin commented:
Great work @dannygoldstein 🎉🚀
